File: OpenBurnBarCore/Sources/OpenBurnBarUI/Views/Recap/RecapTheme.swift
import SwiftUI
import OpenBurnBarInsights
import OpenBurnBarRecap

/// Recap-specific type and colour, layered on `UnifiedDesignSystem` rather than
/// beside it.
///
/// Only two things are genuinely new here: a display scale above `displayLarge`
/// (which the shared system reserves "for hero moments" — this is that moment),
/// and a rule for picking one accent per card from the card's own subject, so
/// colour carries meaning instead of decoration.
public enum RecapTheme {

    // MARK: - Type

    public enum Typography {
        /// The oversized number on a hero card. Monospaced digits so a count-up
        /// animation cannot make the layout jitter.
        public static func expressive(_ size: CGFloat) -> Font {
            .system(size: size, weight: .bold, design: .rounded).monospacedDigit()
        }

        public static let heroNumeral = expressive(72)
        public static let largeNumeral = expressive(48)
        public static let tileNumeral = expressive(34)

        public static let heroHeadline = Font.system(size: 30, weight: .bold, design: .rounded)
        public static let cardHeadline = Font.system(size: 19, weight: .semibold, design: .rounded)
        public static let cardBody = Font.system(size: 15, weight: .regular, design: .rounded)
        /// All-caps eyebrow above a card's headline.
        public static let eyebrow = Font.system(size: 11, weight: .bold, design: .rounded)
        public static let caption = Font.system(size: 13, weight: .medium, design: .rounded)
    }

    // MARK: - Colour

    /// One accent per card, taken from what the card is about.
    ///
    /// A model card is tinted by that model, a provider card by that provider.
    /// Cards with no subject fall back to a tone-derived accent, which keeps the
    /// deck varied without ever reaching for a random palette.
    public static func accent(for card: RecapCard) -> Color {
        if let seed = colorSeed(in: card) {
            return UnifiedDesignSystem.Colors.colorForModel(seed)
        }
        return accent(for: card.tone, kind: card.kind)
    }

    public static func accent(for tone: RecapTone, kind: RecapInsightKind) -> Color {
        switch kind {
        case .record, .milestone:
            return UnifiedDesignSystem.Colors.amber
        case .anomaly:
            return UnifiedDesignSystem.Colors.blaze
        case .funFact:
            return UnifiedDesignSystem.Colors.whimsy
        case .trend, .comparison:
            return UnifiedDesignSystem.Colors.ember
        case .personality:
            return tone == .playful
                ? UnifiedDesignSystem.Colors.whimsy
                : UnifiedDesignSystem.Colors.ember
        }
    }

    private static func colorSeed(in card: RecapCard) -> String? {
        if case let .ranked(entries) = card.visualData, let first = entries.first {
            return first.colorSeed ?? first.label
        }
        // Fleet cards name their subject in the family key.
        let family = card.candidate.family
        if family.hasPrefix("model:") {
            return String(family.dropFirst("model:".count))
        }
        return nil
    }

    /// A restrained two-stop wash for a card's background. Never a rainbow, and
    /// never strong enough to fight the text sitting on it.
    public static func wash(_ accent: Color) -> LinearGradient {
        LinearGradient(
            colors: [accent.opacity(0.14), accent.opacity(0.04)],
            startPoint: .topLeading,
            endPoint: .bottomTrailing
        )
    }

    /// Gradient for expressive numerals — one hue, varied in weight.
    public static func numeralFill(_ accent: Color) -> LinearGradient {
        LinearGradient(
            colors: [accent, accent.opacity(0.72)],
            startPoint: .top,
            endPoint: .bottom
        )
    }

    // MARK: - Metrics

    public enum Layout {
        public static let cardCornerRadius: CGFloat = 20
        public static let heroCornerRadius: CGFloat = 26
        /// Height of one deck row unit. `RecapCardSize.heightUnits` multiplies it.
        public static let rowUnit: CGFloat = 150
        public static let cardPadding: CGFloat = 20
        public static let heroPadding: CGFloat = 28
    }
}

File: android/app/src/main/java/com/openburnbar/ui/recap/RecapTheme.kt
package com.openburnbar.ui.recap

import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.openburnbar.data.recap.RecapCard
import com.openburnbar.data.recap.RecapInsightKind
import com.openburnbar.data.recap.RecapTone
import com.openburnbar.data.recap.RecapVisualData
import com.openburnbar.ui.theme.AuroraColors

object RecapTheme {

    object Typography {
        val heroNumeral = TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Bold,
            fontSize = 72.sp,
            lineHeight = 76.sp,
            letterSpacing = (-1.0).sp,
        )
        val largeNumeral = TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Bold,
            fontSize = 48.sp,
            lineHeight = 52.sp,
            letterSpacing = (-0.5).sp,
        )
        val tileNumeral = TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Bold,
            fontSize = 34.sp,
            lineHeight = 38.sp,
        )
        val heroHeadline = TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Bold,
            fontSize = 28.sp,
            lineHeight = 34.sp,
        )
        val cardHeadline = TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.SemiBold,
            fontSize = 19.sp,
            lineHeight = 24.sp,
        )
        val cardBody = TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Normal,
            fontSize = 15.sp,
            lineHeight = 21.sp,
        )
        val eyebrow = TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Bold,
            fontSize = 11.sp,
            lineHeight = 14.sp,
            letterSpacing = 1.2.sp,
        )
        val caption = TextStyle(
            fontFamily = FontFamily.SansSerif,
            fontWeight = FontWeight.Medium,
            fontSize = 13.sp,
            lineHeight = 18.sp,
        )
    }

    object Layout {
        val cardCornerRadius = 20.dp
        val heroCornerRadius = 26.dp
        val cardPadding = 20.dp
        val heroPadding = 24.dp
        val gutter = 16.dp
    }

    fun accentFor(card: RecapCard): Color {
        val visualData = card.visualData
        if (visualData is RecapVisualData.Ranked) {
            val first = visualData.entries.firstOrNull()
            if (first?.colorSeed != null) {
                return colorForModelSeed(first.colorSeed)
            }
        }
        val family = card.candidate.family
        if (family.startsWith("model:")) {
            return colorForModelSeed(family.removePrefix("model:"))
        }
        return accentFor(card.tone, card.kind)
    }

    fun accentFor(tone: RecapTone, kind: RecapInsightKind): Color {
        return when (kind) {
            RecapInsightKind.RECORD, RecapInsightKind.MILESTONE -> AuroraColors.amber
            RecapInsightKind.ANOMALY -> AuroraColors.blaze
            RecapInsightKind.FUN_FACT -> AuroraColors.whimsy
            RecapInsightKind.TREND, RecapInsightKind.COMPARISON -> AuroraColors.ember
            RecapInsightKind.PERSONALITY -> if (tone == RecapTone.PLAYFUL) AuroraColors.whimsy else AuroraColors.ember
        }
    }

    fun colorForModelSeed(seed: String): Color {
        val s = seed.lowercase()
        return when {
            s.contains("claude") || s.contains("anthropic") -> Color(0xFFD97706)
            s.contains("gpt") || s.contains("openai") -> Color(0xFF10B981)
            s.contains("gemini") || s.contains("google") -> Color(0xFF3B82F6)
            s.contains("grok") || s.contains("xai") -> Color(0xFF8B5CF6)
            s.contains("deepseek") -> Color(0xFF06B6D4)
            s.contains("llama") || s.contains("meta") -> Color(0xFFEC4899)
            else -> AuroraColors.ember
        }
    }

    fun wash(accent: Color): Brush {
        return Brush.linearGradient(
            colors = listOf(accent.copy(alpha = 0.14f), accent.copy(alpha = 0.04f)),
        )
    }

    fun numeralFill(accent: Color): Brush {
        return Brush.verticalGradient(
            colors = listOf(accent, accent.copy(alpha = 0.72f)),
        )
    }
}
